### PR TITLE
Long running spans - pending trace track all spans approach

### PR DESCRIPTION
**Pull request by raphaelgavache**
## Overview

This PR adds support for the long running spans feature, which enables periodic snapshots of running spans to be written. It also includes changes to the PendingTrace and PendingTraceBuffer classes to track long running traces and trigger periodic writes.

## Long running spans spec

Long running spans are tracked
- priority > 0
- while they last less then 12H

Protocol
- write the span with all tags
- span.Duration = write time
- if running span.Metric["_dd.partial_version"] is set. The highest version is the most recent one
- if previously running a complete span will have span.Metric["_dd.was_longrunning"] = 1

Configuration
- feature is disabled by default
- periodic flushes in [20s, 7.5min] with a default of 5min

### Feature disabled
- additional if statement at span initialisation and termination with `pendingTraceBuffer.runningSpansEnabled()`
- no-op on the rest

#### all traces
- PendingTrace keeps track of all spans including running ones
- On write we iterate on all spans including running ones
- PendingTrace is referenced up to 1 second
- One insertion to the PendingTraceBuffer queue per trace
- Running spans are also written if the trace lasts longer then the configured runningFlushInterval (default 5 mins, hard coded min 20s)

#### other
PendingTraceBuffer is waked up every 1s

## Breakdown of the required:
- track running spans
- periodically trigger writes
- encode new tags on spans
- avoid computing stats on running spans

### Track running spans (PendingTrace)

Writes are triggered by PendingTraces and the PendingTraceBuffer. They only point to completed spans.

```
current links:
span ---> pendingTrace ---> completedSpans
                     |
                     ------->  trigger writes
```
New approach when the feature is enabled the pendingTrace will track all spans including runningSpans.
To avoid additional allocations all spans are stored in the `pendingTrace.spans` list.

```
new links
span -> pendingTrace -> allSpans (running, completed)
                     |
                     ------->  trigger writes
```

### Periodic write triggers (PendingTraceBuffer)

The PendingTraceBuffer is an existing thread that can trigger partial writes. It listens to a queue where PendingTraces that meet certain conditions are inserted.
This makes it the perfect place to track long running traces and trigger periodic traces. 

A 1 second periodic wakeup is added to the PendingTraceBuffer thread. On wakeup runningTraces are re-evaluated: a write can be triggered, and traces already completed are cleaned up.

The runningTraces are stored in a dedicated array `LongRunningTracesTracker` with a few helpers.

A LongRunningState has been added to the PendingTrace to ensure that the changes of state happen only once. Exemples
- PendingTrace is sent only once to the PendingTraceBuffer to be tracked
- priority is evaluated only once
- the LongRunningTracker does not contain duplicate PendingTrace
```
PendingTrace state changes

                                  priority <= 0
                    UNDEFINED  --------------------------> NOT_TRACKED
                         |
                         |   priority > 0
                         v
                    TO_TRACK 
                         |
                         |   added to the LongRunningTraceTracker by PendingTraceBuffer
                         v
                    TRACKED 
                         |
                         |   agent feature enabled AND trace is older than flush period
                         v
                    WRITE_RUNNING   //writes will now write running spans
                         |
                         |   trace longer then 12h  
                         v
                    EXPIRED
                      
```

### Encode new tags on spans

The 4 following constraints made this particularly challenging:
- a public `CoreSpan` interface is passed through tracer write logic and we must not add methods on the public interface
- span.Duration == 0 is used in the tracer to know if a span is finished
- long running span tags [should be added in the transport layer](https://github.com/DataDog/dd-trace-java/pull/3452#discussion_r830271996)
- spans should not be cloned

The proposed solution to meet the constraint is:
1. add a version on DDSpan and pass it through the `MetadataConsumer`. The MetadataConsumer is used in the v04 and v05 encoders to set tags.
2. fetch the write time from the PendingTrace if duration == 0 at write time. This unfortunately requires to cast CoreSpan to DDSpan. Check `getDurationNano(CoreSpan<?> span`. Given that the flushed running spans volume is low, this is acceptable.

### Avoid computing stats

A helper `shouldComputeMetrics` is added and running spans are filtered out by checking the span duratio. Note `span.isFinished()` is not available on the CoresSpan public interface, that's why `span.getDuration() == 0` is used.

### CODE REVIEW

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 1]
```diff
@@ -290,6 +354,63 @@ private int write(boolean isPartial) {
...
+          int size = size() + pendingReferenceCount;
+          // If we get here and size is below 0, then the writer before us wrote out at least one
+          // more trace than the size it had when it started. Those span(s) had been added to
+          // finishedSpans by some other thread(s) while the existing spans were being written, but
+          // the completedSpanCount has not yet been incremented. This means that eventually the
+          // count(s) will be incremented, and any new spans added during the period that the count
+          // was negative will be written by someone even if we don't write them right now.
+          if (size > 0 && (!isPartial || size > tracer.getPartialFlushMinSpans())) {
+            trace = new ArrayList<>(size);
+            DDSpan span = spans.pollFirst();
```
Comment (raphaelgavache) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:378]: the function is mostly a copy of the current write span function, with special handling for running spans

File: dd-trace-core/src/main/java/datadog/trace/core/CoreSpan.java [hunk 2]
```diff
@@ -58,6 +58,8 @@
 
   boolean hasSamplingPriority();
 
+  boolean shouldComputeMetric();
```
Comment (dougqh) [inline dd-trace-core/src/main/java/datadog/trace/core/CoreSpan.java:61]: Why are we adding to the public API?
What is the method for?

File: dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java [hunk 3]
```diff
@@ -672,6 +677,11 @@ public boolean isMeasured() {
     return context.isMeasured();
   }
 
+  @Override
+  public boolean shouldComputeMetric() {
+    return (context.isMeasured() || context.isTopLevel()) && longRunningVersion <= 0;
```
Comment (dougqh) [inline dd-trace-core/src/main/java/datadog/trace/core/DDSpan.java:682]: I'd rather not have this exposed through the public API.
This would probably better exist as a static helper method -- not something that end users can call on DDSpan.

File: dd-trace-core/src/main/java/datadog/trace/core/LongPendingTracesTracker.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
+package datadog.trace.core;
+
+import datadog.trace.api.Config;
+import java.util.concurrent.TimeUnit;
+
+public class LongPendingTracesTracker {
```
Comment (dougqh) [inline dd-trace-core/src/main/java/datadog/trace/core/LongPendingTracesTracker.java:6]: I don't see why we needed a separate tracker.  Please explain why this is needed.

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 5]
```diff
@@ -72,7 +72,8 @@ PendingTrace create(@Nonnull DDTraceId traceId) {
   private final TimeSource timeSource;
   private final boolean strictTraceWrites;
   private final HealthMetrics healthMetrics;
-  private final ConcurrentLinkedDeque<DDSpan> finishedSpans = new ConcurrentLinkedDeque<>();
+  private final ConcurrentLinkedDeque<DDSpan> finishedSpans;
```
Comment (dougqh) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:75]: We shouldn't need two collections.  I think we can probably just use finishedSpans are rename it to spans.

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 6]
```diff
@@ -290,6 +354,63 @@ private int write(boolean isPartial) {
     return 0;
   }
 
+  private int writeAllSpans(boolean isPartial) {
```
Comment (dougqh) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:357]: We shouldn't need 2 collections or 2 methods for this.  We're really just changing when we publish from start to finish.  No, need to duplicate a bunch of code and make this harder to maintain.

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java [hunk 7]
```diff
@@ -147,8 +154,22 @@ public void run() {
...
+              runningTraces.flushAndCompact(System.currentTimeMillis());
+            } else {
+              pendingTrace = queue.take(); // block until available;
+            }
 
+            if (pendingTrace == null) {
+              continue;
+            }
+            if (runningSpansEnabled()) {
+              if (pendingTrace instanceof PendingTrace) {
```
Comment (dougqh) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java:169]: The casting here is rather ugly.  I see that was already here, but it is making this hard to follow.

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java [hunk 8]
```diff
@@ -172,9 +193,11 @@ public void run() {
               // Trace has been unmodified long enough, go ahead and write whatever is finished.
               pendingTrace.write();
             } else {
-              // Trace is too new.  Requeue it and sleep to avoid a hot loop.
-              enqueue(pendingTrace);
-              Thread.sleep(SLEEP_TIME_MS);
+              if (!runningSpansEnabled()) {
```
Comment (dougqh) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java:196]: Putting this under runningSpansEnabled check seems off, but maybe, that's just because the code copying ???

File: dd-trace-core/src/main/java/datadog/trace/core/LongPendingTracesTracker.java (hunk 4 above)
Comment (raphaelgavache) [reply dd-trace-core/src/main/java/datadog/trace/core/LongPendingTracesTracker.java]: I replaced it with using the pendingTraceBuffer. I added a separate tracker to allow continuous eviction of completed pendingTrace. By using the pendingTraceBuffer queue and sleeps, pendingTraces that are already flushed are taking the place of long running traces.
With the sleep time of 70ms, and the 4k slots in the array, it takes 286s to pass through the complete queue.

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 9]
```diff
@@ -270,20 +318,38 @@ private int write(boolean isPartial) {
...
+                break;
+              }
+              if (!span.isFinished()) {
+                if (!runningSpanSeen) {
+                  runningSpanSeen = true;
+                  firstRunningSpanID = span.getSpanId();
+                }
+                span.setLongRunningVersion((int) (nowNano / 1000000));
+                trace.add(span);
+                spans.add(span);
```
Comment (raphaelgavache) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:337]: todo: add unfinished span only if priority > 0

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java [hunk 10]
```diff
@@ -183,10 +194,27 @@ public void run() {
...
+      }
+      PendingTrace pendingTrace = (PendingTrace) e;
+      pendingTrace.setEnqueued(false);
+      if (!pendingTrace.hasSpans()) {
+        return;
+      }
+      if ((pendingTrace.getCurrentTimeNano() - pendingTrace.getLastWriteTime()) > flushPeriodNano) {
+        pendingTrace.write();
+      }
+      enqueue(pendingTrace);
```
Comment (raphaelgavache) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java:216]: todo: move back to a special container to handle long running traces. If the pending array gets full, a previously tracked long running trace will stop being tracked

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java [hunk 11]
```diff
@@ -15,6 +17,10 @@
 public abstract class PendingTraceBuffer implements AutoCloseable {
   private static final int BUFFER_SIZE = 1 << 12; // 4096
 
+  public boolean runningSpansEnabled() {
```
Comment (mcculls) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java:20]: `longRunningSpansEnabled()` ?

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java [hunk 12]
```diff
@@ -147,7 +159,17 @@ public void run() {
...
+            if (runningSpansEnabled()) {
+              pendingTrace = queue.poll(1, TimeUnit.SECONDS);
+              runningTracesTracker.flushAndCompact(System.currentTimeMillis());
+            } else {
+              pendingTrace = queue.take(); // block until available;
+            }
+
+            if (pendingTrace == null) {
+              continue;
+            }
```
Comment (mcculls) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java:172]: This check can be moved into the `if (runningSpansEnabled()) {` block as `queue.take()` guarantees that `pendingTrace` will be non-`null` (or the call gets interrupted, in which case an exception will be raised breaking out of the `while` loop)

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 13]
```diff
@@ -169,10 +185,53 @@ void registerSpan(final DDSpan span) {
...
+    // LongRunningTracesTracker.
+    pendingTraceBuffer.enqueue(this);
+  }
+
+  private Integer evaluateSamplingPriority(final DDSpan span) {
+    Integer prio = span.getSamplingPriority();
+    if (prio == null) {
+      DDSpan rootSpan = span.getLocalRootSpan();
+      DDSpan spanToSample = rootSpan == null ? span : rootSpan;
+      tracer.sampler.sample(spanToSample);
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:217]: It feels like this would mess up sampling rates since we are now forcing a sampling decision to span creation time instead of at span write time, not to mention that the a lot of the relevant information to make the sampling decision have maybe not gotten their final form. Is there some way we could delay this to _write time_?

Comment (bantonsson) [reply dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:217]: With the upcoming changes to sampling rules there is the possibility to match on both resources as well as tags, which are both likely to change after span creation. Delaying the decision is important for that to be of any use at all. The tracking can be optimized later when we also start tracking _dirty_ spans for lazy sampling evaluation.

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 14]
```diff
@@ -315,9 +407,39 @@ public long getEndToEndStartTime() {
...
+
   @Override
   public boolean setEnqueued(boolean enqueued) {
     int expected = enqueued ? 0 : 1;
     return IS_ENQUEUED.compareAndSet(this, expected, 1 - expected);
   }
+
+  public boolean writeOnBufferFull() {
+    return !compareAndSetLongRunningState(
+        LongRunningTracesTracker.TO_TRACK, LongRunningTracesTracker.NOT_TRACKED);
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:430]: This logic is a bit hard to follow, and I need to reread it when I see it, so a comment would be useful.

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 15]
```diff
@@ -75,8 +80,7 @@ PendingTrace create(@Nonnull DDTraceId traceId) {
   private final HealthMetrics healthMetrics;
   private final TraceConfig traceConfig;
 
-  private final ConcurrentLinkedDeque<DDSpan> finishedSpans = new ConcurrentLinkedDeque<>();
-
+  private final ConcurrentLinkedDeque<DDSpan> spans;
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:83]: A comment about the _dual_ roles of `span` would be good.

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 16]
```diff
@@ -75,8 +80,7 @@ PendingTrace create(@Nonnull DDTraceId traceId) {
...
   // We must maintain a separate count because ConcurrentLinkedDeque.size() is a linear operation.
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:84]: Update comment to include long running use case also.

**Review by bantonsson — APPROVED** [review]
Only minor nitpicks.

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java [hunk 17]
```diff
@@ -147,7 +171,16 @@ public void run() {
         try {
           while (!closed && !Thread.currentThread().isInterrupted()) {
 
-            Element pendingTrace = queue.take(); // block until available.
+            Element pendingTrace = null;
+            if (longRunningSpansEnabled()) {
+              pendingTrace = queue.poll(1, TimeUnit.SECONDS);
+              runningTracesTracker.flushAndCompact(System.currentTimeMillis());
```
Comment (bantonsson) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTraceBuffer.java:177]: This should be `timeSource.getCurrentTimeMillis()`

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java (hunk 16 above)
Comment (raphaelgavache) [reply dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java]: Nice catch, this comment does not make sense anymore. I just removed it as I find the name explicit. I could not find also a relevant simple comment to explain the uses of counting completed spans

File: internal-api/src/main/java/datadog/trace/api/Config.java [hunk 18]
```diff
@@ -1528,6 +1550,14 @@ public boolean isTraceEnabled() {
     return instrumenterConfig.isTraceEnabled();
   }
 
+  public boolean isLongRunningTracesEnabled() {
+    return longRunningTraceEnabled;
+  }
+
+  public long getLongRunningFlushInterval() {
+    return longRunningTraceFlushInterval;
```
Comment (mcculls) [inline internal-api/src/main/java/datadog/trace/api/Config.java:1558]: could you update the field / method names to be consistent - ie. `foo` and `getFoo()`, `bar` and `isBar()` - thanks

File: internal-api/src/main/java/datadog/trace/api/Config.java [hunk 19]
```diff
@@ -1464,6 +1467,25 @@ private Config(final ConfigProvider configProvider, final InstrumenterConfig ins
...
     apiKey = tmpApiKey;
 
+    boolean longRunningEnabled =
+        configProvider.getBoolean(
+            TracerConfig.TRACE_LONG_RUNNING_ENABLED,
+            ConfigDefaults.DEFAULT_LONG_RUNNING_TRACE_ENABLED);
+    this.longRunningTraceFlushInterval =
+        configProvider.getLong(
+            TracerConfig.TRACE_LONG_RUNNING_FLUSH_INTERVAL,
+            ConfigDefaults.DEFAULT_LONG_RUNNING_TRACE_FLUSH_INTERVAL);
```
Comment (mcculls) [inline internal-api/src/main/java/datadog/trace/api/Config.java:1477]: could you also make these consistent, ie. name the default as `DEFAULT_XYZ` for a flag called `XYZ`

File: internal-api/src/main/java/datadog/trace/api/Config.java [hunk 20]
```diff
@@ -1464,6 +1467,25 @@ private Config(final ConfigProvider configProvider, final InstrumenterConfig ins
...
+
+    if (longRunningEnabled
+        && (longRunningTraceFlushInterval < 20 || longRunningTraceFlushInterval > 450)) {
+      log.warn(
+          "Provided long running trace flush interval of {} seconds. It should be between 20 seconds and 7.5 minutes. T"
+              + "he feature is disabled.",
```
Comment (mcculls) [inline internal-api/src/main/java/datadog/trace/api/Config.java:1483]: splitting the string between the sentences is more readable, rather than splitting mid-word

Comment (mcculls) [reply internal-api/src/main/java/datadog/trace/api/Config.java:1483]: Also would it be better to move the flush interval inside the allowed range and keep it enabled? (Can still warn that the original value was outside the accepted range and has been adjusted.)

File: dd-trace-core/src/main/java/datadog/trace/core/monitor/TracerHealthMetrics.java [hunk 21]
```diff
@@ -104,6 +104,12 @@ public class TracerHealthMetrics extends HealthMetrics implements AutoCloseable
       CountersFactory.createFixedSizeStripedCounter(8);
   private final FixedSizeStripedLongCounter clientSpansWithoutContext =
       CountersFactory.createFixedSizeStripedCounter(8);
+  private final FixedSizeStripedLongCounter longRunningTracesWrite =
```
Comment (mcculls) [inline dd-trace-core/src/main/java/datadog/trace/core/monitor/TracerHealthMetrics.java:107]: `longRunningTracesWrites` to match the other new metric names?

File: dd-trace-core/src/main/java/datadog/trace/core/monitor/TracerHealthMetrics.java [hunk 22]
```diff
@@ -104,6 +104,12 @@ public class TracerHealthMetrics extends HealthMetrics implements AutoCloseable
...
+      CountersFactory.createFixedSizeStripedCounter(8);
+  private final FixedSizeStripedLongCounter longRunningTracesMisses =
+      CountersFactory.createFixedSizeStripedCounter(8);
+  private final FixedSizeStripedLongCounter longRunningTracesExpires =
+      CountersFactory.createFixedSizeStripedCounter(8);
```
Comment (mcculls) [inline dd-trace-core/src/main/java/datadog/trace/core/monitor/TracerHealthMetrics.java:112]: should be `longRunningTracesDropped` and `longRunningTracesExpired` to match the actual metrics names

File: dd-trace-core/src/main/java/datadog/trace/core/monitor/TracerHealthMetrics.java [hunk 23]
```diff
@@ -308,6 +314,13 @@ public void onFailedSend(
     onSendAttempt(traceCount, sizeInBytes, response);
   }
 
+  @Override
+  public void onLongRunningUpdate(final int missedAdd, final int writes, final int expires) {
```
Comment (mcculls) [inline dd-trace-core/src/main/java/datadog/trace/core/monitor/TracerHealthMetrics.java:318]: `dropped, write, expired` to match the actual metrics names?

File: internal-api/src/main/java/datadog/trace/api/Config.java [hunk 24]
```diff
@@ -1528,6 +1550,14 @@ public boolean isTraceEnabled() {
     return instrumenterConfig.isTraceEnabled();
   }
 
+  public boolean isLongRunningTracesEnabled() {
+    return longRunningTraceEnabled;
```
Comment (mcculls) [inline internal-api/src/main/java/datadog/trace/api/Config.java:1554]: Consider adding this flag to the `toString()` method as it will help support identify users that have this enabled

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 25]
```diff
@@ -315,9 +411,45 @@ public long getEndToEndStartTime() {
...
+   *
+   * <p>If the pendingTrace is not sent to the LongRunningTracesTracker, it will be immediately
+   * written. Otherwise, the pendingTrace won't be tracked and no write is triggered.
+   */
+  public boolean writeOnBufferFull() {
+    return !compareAndSetLongRunningState(
+        LongRunningTracesTracker.TO_TRACK, LongRunningTracesTracker.NOT_TRACKED);
+  }
+
+  public static long getDurationNano(CoreSpan<?> span) {
```
Comment (mcculls) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:443]: At the moment this method is only called in the trace mappers - could we add a comment explaining when this should be used instead of `span.getDurationNano()` ?

File: dd-trace-core/src/main/java/datadog/trace/core/LongRunningTracesTracker.java [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+  private final HealthMetrics healthMetrics;
+  private long lastFlushMilli = 0;
+
+  private final int maxTrackedTraces;
+  private final int flushPeriodMilli;
+  private final long maxTrackedDurationMilli = TimeUnit.HOURS.toMillis(12);
+  private final List<PendingTrace> traceArray = new ArrayList<>(1 << 4);
+  private int missedAdd = 0;
+  private int writes = 0;
+  private int expires = 0;
```
Comment (mcculls) [inline dd-trace-core/src/main/java/datadog/trace/core/LongRunningTracesTracker.java:22]: `dropped, write, expired` to match actual metric names for consistency?

File: dd-trace-core/src/main/java/datadog/trace/core/LongRunningTracesTracker.java [hunk 27]
```diff
@@ -0,0 +1,133 @@
...
+    // PendingTraces are added only once
+    if (!trace.compareAndSetLongRunningState(TO_TRACK, TRACKED)) {
+      return false;
+    }
+    this.addTrace(trace);
+    return true;
+  }
+
+  private void addTrace(PendingTrace trace) {
+    if (trace == null || trace.empty()) {
```
Comment (mcculls) [inline dd-trace-core/src/main/java/datadog/trace/core/LongRunningTracesTracker.java:56]: AFAICT this is only called by `add` which guarantees it will only pass non-`null` traces so the `null` check can be removed here

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java [hunk 28]
```diff
@@ -279,26 +344,57 @@ private int write(boolean isPartial) {
...
     }
     return 0;
   }
 
+  public int enqueueSpansToWrite(List<DDSpan> trace, boolean writeRunningSpans) {
+    int completedSpans = 0;
+    boolean runningSpanSeen = false;
+    long firstRunningSpanID = 0;
+    long nowNano = getCurrentTimeNano();
+    setLastWriteTime(nowNano);
```
Comment (mcculls) [inline dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java:368]: previously we just added the finished spans to the trace - now we also get the current time and write it to a `volatile` even when long-running traces are not enabled

is there a way we can keep the simpler loop (and skip the time check and `volatile` write) when we know we aren't capturing long-running traces?

**Review by mcculls — APPROVED** [review]
Mostly comments about naming consistency, plus a question whether we can keep the simpler enqueuing of finished spans when long-running traces are not enabled - but no blockers

File: dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java (hunk 28 above)
Comment (raphaelgavache) [reply dd-trace-core/src/main/java/datadog/trace/core/PendingTrace.java]: Nice catch, this was an easy to removed and it simplified PendingTraceBuffer tests

File: dd-trace-core/src/main/java/datadog/trace/core/LongRunningTracesTracker.java (hunk 27 above)
Comment (raphaelgavache) [reply dd-trace-core/src/main/java/datadog/trace/core/LongRunningTracesTracker.java]: yes, it passes the null insertion test once removed

File: internal-api/src/main/java/datadog/trace/api/Config.java [hunk 29]
```diff
@@ -3411,6 +3414,10 @@ public String toString() {
         + cwsEnabled
         + ", cwsTlsRefresh="
         + cwsTlsRefresh
+        + ", longRunningTracesEnabled="
+        + longRunningTraceEnabled
+        + ", longRunningTraceFlushInterval="
```
Comment (bantonsson) [inline internal-api/src/main/java/datadog/trace/api/Config.java:3463]: Mixing `traces` and `trace`..
